File: dotnet-maui/DittoMauiTasksApp/Data/DittoService.cs
using DittoSDK;
using Microsoft.Extensions.Logging;
using System.Text.Json;
using System.Collections.ObjectModel;
using DittoMauiTasksApp.Model;

namespace DittoMauiTasksApp.Data;

/// <summary>
/// Manages Ditto database operations for the tasks application, including CRUD operations,
/// data synchronization, and real-time data observation. This class serves as the primary
/// data access layer for task management functionality.
/// </summary>

public class DittoService(ILogger<DittoService> logger)
    : IDataService
{
    private readonly ILogger<DittoService> _logger = logger;
    private const string SelectQuery = "SELECT * FROM tasks WHERE NOT deleted";
    
    /// <summary>
    /// Lock object used to ensure thread-safe initialization.
    /// Only one thread can execute the initialization logic at a time.
    /// </summary>
    private static readonly SemaphoreSlim InitializationLock = new SemaphoreSlim(1, 1);
    
    /// <summary>
    /// Flag indicating whether initialization has already been completed.
    /// This prevents redundant initialization attempts.
    /// </summary>
    private static volatile bool _isInitialized = false;

#nullable enable
    private Ditto? _ditto = null;
    private DittoStoreObserver? _dittoStoreObserver = null;
    private DittoSyncSubscription? _syncSubscription = null;

    /// <summary>
    /// DittoDiffer instance used to calculate the delta changes between syncs.
    /// This enables efficient UI updates by only processing changed items.
    /// </summary>
    /// <see href="https://docs.ditto.live/sdk/latest/crud/read#diffing-results"/>
    private readonly DittoDiffer _dittoDiffer = new DittoDiffer();
    
    /// <summary>
    /// Cache of the previous query results for diffing operations.
    /// </summary>
    private List<DittoQueryResultItem> _previousItems = [];

    /// <summary>
    /// Initializes the Ditto instance with the specified configuration and data path.
    /// This method is thread-safe and ensures initialization only occurs once, even if
    /// called concurrently from multiple threads. Subsequent calls will return immediately
    /// if initialization has already been completed.
    /// </summary>
    /// <param name="dittoConfig">The Ditto configuration containing app ID, tokens, and URLs.</param>
    /// <param name="dataPath">The file system path where Ditto should store its data.</param>
    /// <returns>A task that represents the asynchronous initialization operation.</returns>
    /// <exception cref="Exception">Thrown when Ditto initialization fails.</exception>
    public async Task Initialize(DittoConfig dittoConfig, string dataPath)
    {
        // Quick check to avoid unnecessary lock acquisition if already initialized
        if (_isInitialized)
        {
            _logger.LogDebug("DittoService already initialized, skipping initialization");
            return;
        }
        // Acquire the lock to ensure only one thread can initialize at a time
        await InitializationLock.WaitAsync();
        try
        {
            // Double-check pattern: verify initialization hasn't been completed by another thread
            if (_isInitialized)
            {
                _logger.LogDebug("DittoService already initialized by another thread, skipping initialization");
                return;
            }
            _ditto = new Ditto(DittoIdentity
                .OnlinePlayground(
                    dittoConfig.AppId,
                    dittoConfig.PlaygroundToken,
                    false, // This is required to be set to false, to use the correct URLs with Ditto Portal
                    dittoConfig.AuthUrl), dataPath);

            _ditto.UpdateTransportConfig(config =>
            {
                // Add the websocket URL to the transport configuration.
                config.Connect.WebsocketUrls.Add(dittoConfig.WebsocketUrl);
            });

            // disable sync with v3 peers, required for DQL
            _ditto.DisableSyncWithV3();

            // Disable DQL strict mode
            // https://docs.ditto.live/dql/strict-mode
            await _ditto.Store.ExecuteAsync("ALTER SYSTEM SET DQL_STRICT_MODE = false");
            
            await InsertInitialTasks();
            
            // Register a subscription, which determines what data syncs to this peer
            // https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
            _syncSubscription = _ditto?.Sync.RegisterSubscription(SelectQuery);
            
            // Mark initialization as complete
            _isInitialized = true;
            _logger.LogInformation("DittoService initialization completed successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "DittoService initialization failed: {Message}", ex.Message);
            throw; // Re-throw to allow callers to handle the error
        }
        finally
        {
            // Always release the lock, even if an exception occurs
            InitializationLock.Release();
        }
    }

    /// <summary>
    /// Initializes the Ditto instance with the specified configuration using the default
    /// local application data path for data storage.
    /// </summary>
    /// <param name="dittoConfig">The Ditto configuration containing app ID, tokens, and URLs.</param>
    /// <returns>A task that represents the asynchronous initialization operation.</returns>
    public async Task Initialize(DittoConfig dittoConfig)
    {
        var dataPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "ditto");
        await Initialize(dittoConfig, dataPath);
    }

    /// <summary>
    /// Registers a store observer to monitor changes in the tasks collection and automatically
    /// update the provided ObservableCollection. Uses diffing to efficiently update only
    /// changed items, preventing unnecessary UI repaints.
    /// </summary>
    /// <param name="tasks">The ObservableCollection to be updated when task data changes.</param>
    /// <see href="https://docs.ditto.live/sdk/latest/crud/observing-data-changes#setting-up-store-observers"/>
    /// <see href="https://docs.ditto.live/sdk/latest/crud/read#diffing-results"/>
    public void RegisterObservers(ObservableCollection<DittoTask> tasks)
    {
        // Register an observer, which runs against the local database on this peer
        // https://docs.ditto.live/sdk/latest/crud/observing-data-changes#setting-up-store-observers
        _dittoStoreObserver = _ditto?.Store.RegisterObserver(SelectQuery, (queryResult) =>
        {
            try
            {
                // .NET specifically with ObservableCollection will repaint the entire list if you just clear it and add new items.
                // Diffing the items allows us to only update the changed items, but at the expense of some memory and CPU usage.
                // This is a trade-off that is acceptable for this example, but may lead to problems on larger datasets.
                // https://docs.ditto.live/sdk/latest/crud/read#diffing-results
                var diff = _dittoDiffer.Diff(queryResult.Items);

                // For testing, we'll execute directly since we're not in a UI context
                try
                {
                    //handle deletions
                    foreach (var taskItem in diff.Deletions.Select(index => JsonSerializer
                                 .Deserialize<DittoTask>(_previousItems[index].JsonString())).OfType<DittoTask>())
                    {
                        tasks.Remove(tasks.First(x => x.Id == taskItem.Id));
                    }

                    //handle insertions
                    foreach (var index in diff.Insertions)
                    {
                        var item = queryResult.Items[index];
                        var taskItem = JsonSerializer
                            .Deserialize<DittoTask>(item.JsonString());
                        if (taskItem != null)
                        {
                            tasks.Add(taskItem);
                        }
                    }

                    //handle updates
                    foreach (var index in diff.Updates)
                    {
                        var item = queryResult.Items[index];
                        var taskItem = JsonSerializer
                            .Deserialize<DittoTask>(item.JsonString());
                        if (taskItem != null)
                        {
                            var existingTask = tasks.FirstOrDefault(x => x.Id == taskItem.Id);
                            if (existingTask == null) continue;
                            //implement soft delete for the deleted field
                            if (taskItem.Deleted)
                            {
                                tasks.Remove(existingTask);
                            }
                            else
                            {
                                existingTask.Title = taskItem.Title;
                                existingTask.Done = taskItem.Done;
                            }
                        }
                    }
                }
                catch (Exception e)
                {
                    _logger.LogError($"TasksPageViewModel: Error: Unable to update list view model: {e.Message}");
                }

                _previousItems = queryResult.Items;
            }
            catch (Exception e)
            {
                _logger.LogError($"TasksPageViewModel: Error: Unable to process tasks collection change: {e.Message}");
            }
        });
    }

    /// <summary>
    /// Starts the Ditto synchronization process and registers a subscription for the tasks collection.
    /// This enables real-time data synchronization with other peers and the Ditto cloud.
    /// </summary>
    /// <see href="https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions"/>
    public void StartSync()
    {
        _ditto?.StartSync();
    }

    /// <summary>
    /// Stops the Ditto synchronization process and cancels the active subscription.
    /// This should be called when sync is no longer needed or when the application is shutting down.
    /// </summary>
    public void StopSync()
    {
        _ditto?.StopSync();
        _syncSubscription?.Cancel();
    }

    /// <summary>
    /// Populates the tasks collection with initial seed data if it's empty.
    /// Inserts four predefined tasks using the INITIAL keyword to prevent duplicates.
    /// </summary>
    /// <returns>A task that represents the asynchronous insertion operation.</returns>
    /// <exception cref="Exception">Thrown when the insertion operation fails.</exception>
     public async Task InsertInitialTasks()
     {
        try
        {
            var initialTasks = new List<Dictionary<string, object>>
            {
                new()
                {
                    { "_id", "50191411-4C46-4940-8B72-5F8017A04FA7" },
                    { "title", "Buy groceries" },
                    { "done", false },
                    { "deleted", false }
                },
                new()
                {
                    { "_id", "6DA283DA-8CFE-4526-A6FA-D385089364E5" },
                    { "title", "Clean the kitchen" },
                    { "done", false },
                    { "deleted", false }
                },
                new()
                {
                    { "_id", "5303DDF8-0E72-4FEB-9E82-4B007E5797F0" },
                    { "title", "Schedule dentist appointment" },
                    { "done", false },
                    { "deleted", false }
                },
                new()
                {
                    { "_id", "38411F1B-6B49-4346-90C3-0B16CE97E174" },
                    { "title", "Pay bills" },
                    { "done", false },
                    { "deleted", false }
                }
            };

            const string insertCommand = "INSERT INTO tasks INITIAL DOCUMENTS (:task)";

            if (_ditto != null)
            {
                foreach (var task in initialTasks)
                {
                    await _ditto.Store.ExecuteAsync(insertCommand, new Dictionary<string, object>()
                    {
                        { "task", task }
                    });
                }
            }
        }
        catch (Exception e)
        {
            _logger.LogError($"TasksPageViewModel: Error adding initial tasks: {e.Message}");
        }
    }

    /// <summary>
    /// Adds a new task to the tasks collection in the Ditto store.
    /// </summary>
    /// <param name="document">A dictionary containing the task data with keys: _id, title, done, deleted.</param>
    /// <returns>A task that represents the asynchronous insertion operation.</returns>
    /// <exception cref="Exception">Thrown when the insertion operation fails.</exception>
    public async Task AddTask(Dictionary<string, object> document)
    {
        const string insertCommand = "INSERT INTO tasks DOCUMENTS (:doc)";
        if (_ditto != null)
        {
            await _ditto.Store.ExecuteAsync(insertCommand, new Dictionary<string, object>()
            {
                { "doc", document }
            });
        }
    }

    /// <summary>
    /// Updates the title of an existing task in the Ditto store.
    /// </summary>
    /// <param name="task">The DittoTask object containing the updated title and task ID.</param>
    /// <returns>A task that represents the asynchronous update operation.</returns>
    /// <exception cref="Exception">Thrown when the update operation fails.</exception>
    public async Task EditTask(DittoTask task)
    {
        if (_ditto != null)
        {
            const string updateQuery = "UPDATE tasks SET title = :title WHERE _id = :id";
            await _ditto.Store.ExecuteAsync(updateQuery, new Dictionary<string, object>()
            {
                { "title", task.Title },
                { "id", task.Id }
            });
        }
    }

    /// <summary>
    /// Marks a task as deleted using the soft delete pattern. The task remains in the database
    /// but is marked as deleted and will not appear in normal queries.
    /// </summary>
    /// <param name="task">The DittoTask object to be marked as deleted.</param>
    /// <returns>A task that represents the asynchronous deletion operation.</returns>
    /// <exception cref="Exception">Thrown when the deletion operation fails.</exception>
    public async Task DeleteTask(DittoTask task)
    {
        if (_ditto != null)
        {
            const string updateQuery = "UPDATE tasks SET deleted = true WHERE _id = :id";
            await _ditto.Store.ExecuteAsync(updateQuery, new Dictionary<string, object>()
            {
                { "id", task.Id }
            });
        }
    }

    /// <summary>
    /// Updates the completion status of a task in the Ditto store. Only updates if the
    /// current done state differs from the new state to avoid unnecessary operations.
    /// </summary>
    /// <param name="task">The DittoTask object containing the updated done status and task ID.</param>
    /// <returns>A task that represents the asynchronous update operation.</returns>
    /// <exception cref="Exception">Thrown when the update operation fails.</exception>
    public async Task UpdateTaskDoneAsync(DittoTask task)
    {
        if (_ditto != null)
        {
            var doc = new Dictionary<string, object>
            {
                { "newDoneState", task.Done },
                { "id", task.Id }
            };

            const string updateQuery =
                "UPDATE tasks SET done = :newDoneState WHERE _id = :id AND done != :newDoneState";
            await _ditto.Store.ExecuteAsync(updateQuery, doc);
        }
    }
}

File: swift/Tasks/DittoService.swift
import DittoSwift
import Foundation

@MainActor class DittoStateManager: ObservableObject {
    @Published var isInitialized: Bool = false

    nonisolated init() { }

    func setIsInitialized() {
        self.isInitialized = true
    }

    func setIsNotInitialized() {
        self.isInitialized = false
    }
}


actor DittoService {
    var subscription: DittoSyncSubscription? = nil
    var storeObserver: DittoStoreObserver? = nil
    var ditto: Ditto? = nil

    static var shared = DittoService()

    init() {}

    /// Performs cleanup of Ditto resources
    ///
    /// This method handles the graceful shutdown of Ditto components by:
    /// - Cancelling any active subscriptions
    /// - Cancelling store observers
    /// - Stopping the Ditto sync process
    func deinitialize(dittoStateManager: DittoStateManager) {
        subscription?.cancel()
        subscription = nil
        storeObserver?.cancel()
        storeObserver = nil
        if let dittoInstance = ditto {
            if dittoInstance.isSyncActive {
                dittoInstance.stopSync()
            }
            ditto = nil
        }
        DispatchQueue.main.async {
            dittoStateManager.setIsNotInitialized()
        }
    }

    /// Initializes the Ditto instance and configures its environment.
    ///
    /// - Throws: An error if Ditto configuration or DQL commands fail
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/install-guides/swift#integrating-and-initializing-sync
    /// - SeeAlso: https://docs.ditto.live/dql/strict-mode
    func initializeIfNeeded(dittoStateManager: DittoStateManager) async throws {
        if await dittoStateManager.isInitialized {
            return
        }

        //setup logging level
        let isPreview: Bool =
            ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"]
            == "1"
        if !isPreview {
            DittoLogger.minimumLogLevel = .debug
        }

        // Setup Ditto Identity
        let ditto = Ditto(
            identity: .onlinePlayground(
                appID: Env.DITTO_APP_ID,
                token: Env.DITTO_PLAYGROUND_TOKEN,
                enableDittoCloudSync: false,
                customAuthURL: URL(string: Env.DITTO_AUTH_URL)
            )
        )
        self.ditto = ditto

        // Set the Ditto Websocket URL
        ditto.updateTransportConfig { transportConfig in
            transportConfig.connect.webSocketURLs.insert(
                Env.DITTO_WEBSOCKET_URL
            )
        }
        // disable sync with v3 peers, required for DQL
        try ditto.disableSyncWithV3()

        // Disable DQL strict mode
        try await ditto.store.execute(
            query: "ALTER SYSTEM SET DQL_STRICT_MODE = false"
        )
        try await self.populateTaskCollection()
        try self.registerSubscription()
        DispatchQueue.main.async {
            dittoStateManager.setIsInitialized()
        }
    }

    /// Populates the Ditto tasks collection with initial seed data if it's empty
    ///
    /// This method creates a set of predefined tasks in the Ditto store by:
    /// - Defining an array of initial `TaskModel` objects with unique IDs and titles
    /// - Inserting each task into the Ditto store using DQL (Ditto Query Language)
    /// - Using the INITIAL keyword to only insert if documents don't already exist
    ///
    /// The initial tasks include common todo items like:
    /// - Buy groceries
    /// - Clean the kitchen
    /// - Schedule dentist appointment
    /// - Pay bills
    ///
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/read#using-args-to-query-dynamic-values///
    ///
    /// - Throws: A DittoError if the insert operations fail
    func populateTaskCollection() async throws {

        let initialTasks: [TaskModel] = [
            TaskModel(
                _id: "50191411-4C46-4940-8B72-5F8017A04FA811",
                title: "Buy groceries"
            ),
            TaskModel(
                _id: "6DA283DA-8CFE-4526-A6FA-D385089364E811",
                title: "Clean the kitchen"
            ),
            TaskModel(
                _id: "5303DDF8-0E72-4FEB-9E82-4B007E5797F911",
                title: "Schedule dentist appointment"
            ),
            TaskModel(
                _id: "38411F1B-6B49-4346-90C3-0B16CE97E17911",
                title: "Pay bills"
            ),
        ]
        for task in initialTasks {
            if let ditto {
                // https://docs.ditto.live/sdk/latest/crud/write#inserting-documents
                try await ditto.store.execute(
                    query: "INSERT INTO tasks INITIAL DOCUMENTS (:task)",
                    arguments: [
                        "task":
                            [
                                "_id": task._id,
                                "title": task.title,
                                "done": task.done,
                                "deleted": task.deleted,
                            ]
                    ]
                )
            } else {
                throw ManagerError.dittoNotInitialized(
                    "Can't INSERT INTIIAL DOCUMENTS - Ditto is not initialized"
                )
            }
        }
    }

    /// Registers observers for the planets collection to handle real-time updates.
    ///
    /// This method sets up a live query observer that:
    /// - Monitors the tasks collection for changes
    /// - Updates the @Published tasks array when changes occur
    /// - Filters out tasks NOT deleted
    ///
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/read
    ///
    /// - Throws: A DittoError if the observer cannot be registered
    func registerObservers(updateTaskModels: @escaping ([TaskModel]?) async -> Void) throws {
        if let ditto {
            let observerQuery = "SELECT * FROM tasks WHERE NOT deleted"
            storeObserver = try ditto.store.registerObserver(
                query: observerQuery
            ) { results in
                Task { @MainActor in
                    // Create new TaskModel instances and update the published property
                    let tasks = results.items.compactMap {
                        TaskModel($0.jsonString())
                    }
                    // Call the callback on the main actor
                    await updateTaskModels(tasks)
                }
            }
        } else {
            throw ManagerError.dittoNotInitialized(
                "Can't register observer - Ditto is not initialized"
            )
        }
    }

    /// Registers a subscription to sync all tasks across peers.
    ///
    /// This method sets up a sync subscription that:
    /// - Creates a subscription for the entire tasks collection
    /// - Enables data to be synced between devices when sync is active
    /// - Stores the subscription reference for later cancellation
    ///
    /// The subscription remains active until explicitly cancelled or when sync is stopped.
    ///
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
    ///
    /// - Throws: A DittoError if the subscription registration fails
    func registerSubscription() throws {
        if let ditto {
            // https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
            let subscriptionQuery = "SELECT * from tasks"
            subscription = try ditto.sync.registerSubscription(
                query: subscriptionQuery
            )
        } else {
            throw ManagerError.dittoNotInitialized(
                "Can't register subscription - Ditto is not initialized"
            )
        }
    }

    /// Enables or disables Ditto sync based on the provided value.
    ///
    /// This method checks the current sync state and starts or stops
    /// Ditto sync accordingly:
    /// - If `newValue` is `true` and sync is not active, it starts sync.
    /// - If `newValue` is `false` and sync is active, it stops sync.
    ///
    /// - Parameter newValue: A Boolean indicating whether sync should be enabled (`true`) or disabled (`false`).
    /// - Throws: A DittoError if the operation fails
    func setSyncEnabled(_ newValue: Bool) throws {
        if let ditto {
            if !ditto.isSyncActive && newValue {
                try startSync()
            } else if ditto.isSyncActive && !newValue {
                stopSync()
            }
        }
    }

    /// Starts Ditto sync and registers a subscription for the tasks collection.
    ///
    /// This method:
    /// - Initiates the Ditto sync process if the Ditto instance is available.
    ///
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/install-guides/swift#integrating-and-initializing-sync
    /// - Throws: A DittoError if the operation fails
    private func startSync() throws {
        // https://docs.ditto.live/sdk/latest/install-guides/swift#integrating-and-initializing-sync
        try ditto?.startSync()
    }

    /// Stops Ditto sync and cancels the active subscription for the tasks collection.
    ///
    /// This method:
    /// - Cancels the current sync subscription to stop receiving updates from remote peers.
    /// - Sets the subscription property to `nil`.
    /// - Stops the Ditto sync process if it is active.
    ///
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/sync/syncing-data#canceling-subscriptions
    /// - Throws: A DittoError if the operation fails
    private func stopSync() {
        // https://docs.ditto.live/sdk/latest/sync/syncing-data#canceling-subscriptions
        ditto?.stopSync()
    }

    /// Creates a new TaskModel document in the Ditto store.
    ///
    /// This method:
    /// - Creates a new document in the tasks collection
    ///
    /// - Parameter task: The TaskModel to add to the store
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/create#creating-documents
    ///
    /// - Throws: A DittoError if the insert operation fails
    func insertTaskModel(_ task: TaskModel) async throws {
        let newTask = task.value
        // https://docs.ditto.live/dql/insert
        // https://docs.ditto.live/sdk/latest/crud/create#creating-documents
        let query = "INSERT INTO tasks DOCUMENTS (:newTask)"
        if let ditto {
            // https://docs.ditto.live/sdk/latest/crud/create#creating-documents
            try await ditto.store.execute(
                query: query,
                arguments: ["newTask": newTask]
            )
        } else {
            throw ManagerError.dittoNotInitialized(
                "Can't INSERT DOCUMENT - Ditto is not initialized"
            )
        }
    }

    /// Updates an existing TaskModel's properties in the Ditto store.
    ///
    /// This method uses DQL to update all mutable fields of the TaskModel
    ///
    /// - Parameter planet: The TaskModel object containing the updated values
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/update#updating
    ///
    /// - Throws: A DittoError if the update operation fails
    func updateTaskModel(_ task: TaskModel) async throws {
        // https://docs.ditto.live/dql/update#basic-update
        // https://docs.ditto.live/sdk/latest/crud/update#updating
        let query = """
            UPDATE tasks SET
            title = :title,
            done = :done,
            deleted = :deleted
            WHERE _id == :_id
            """
        if let ditto {
            // https://docs.ditto.live/sdk/latest/crud/update#updating
            try await ditto.store.execute(
                query: query,
                arguments: [
                    "title": task.title,
                    "done": task.done,
                    "deleted": task.deleted,
                    "_id": task._id,
                ]
            )
        } else {
            throw ManagerError.dittoNotInitialized(
                "Can't UPDATE DOCUMENT - Ditto is not initialized"
            )
        }
    }

    /// Toggles the completion status of a task in the Ditto store
    ///
    /// This method:
    /// - Inverts the current 'done' status of the task
    /// - Updates only the 'done' field in the store using DQL
    /// - Maintains all other task properties unchanged
    ///
    /// - Parameter task: The TaskModel to toggle completion status for
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/update#updating
    ///
    /// - Throws: A DittoError if the update operation fails
    func toggleComplete(task: TaskModel) async throws {
        let done = !task.done
        // https://docs.ditto.live/dql/update#basic-update
        // https://docs.ditto.live/sdk/latest/crud/update#updating
        let query = """
            UPDATE tasks
            SET done = :done 
            WHERE _id == :_id
            """
        if let ditto {
            // https://docs.ditto.live/sdk/latest/crud/update#updating
            try await ditto.store.execute(
                query: query,
                arguments: ["done": done, "_id": task._id]
            )
        } else {
            throw ManagerError.dittoNotInitialized(
                "Can't UPDATE DOCUMENT - Ditto is not initialized"
            )
        }
    }

    /// Delete a TaskModel by setting its deleted flag to true.
    ///
    /// This method implements the 'Soft-Delete' pattern, which:
    /// - Marks the TaskModel as deleted instead of evicting it
    /// - Removes it from active queries and views
    /// - Maintains the data for historical purposes
    ///
    /// - Parameter task: The TaskModel to soft-delet
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern
    ///
    /// - Throws: A DittoError if the archive operation fails
    func deleteTaskModel(_ task: TaskModel) async throws {
        // https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern
        let query = "UPDATE tasks SET deleted = true WHERE _id = :_id"
        if let ditto {

            // https://docs.ditto.live/sdk/latest/crud/update#updating
            try await ditto.store.execute(
                query: query,
                arguments: ["_id": task._id]
            )
        } else {
            throw ManagerError.dittoNotInitialized(
                "Can't SOFT DELETE DOCUMENT - Ditto is not initialized"
            )
        }
    }
}

enum ManagerError: Error {
    case dittoNotInitialized(String)
}
